src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -35,46 +44,395 @@
 public class TenantPermsAPI implements Tenantpermissions {
 
   private static final String PERMISSION_NAME_FIELD = "'permissionName'";
+  private static final String MODULE_NAME_FIELD = "'moduleName'";
+  private static final String DEFINED_FIELD = "'defined'";
   private static final String TABLE_NAME_PERMS = "permissions";
+  private static final String TABLE_NAME_PERMSUSERS = "permissions_users";
+  public static final String DEPRECATED_PREFIX = "(deprecated) ";
 
   private final Logger logger = LoggerFactory.getLogger(TenantPermsAPI.class);
 
   @Override
   public void postTenantpermissions(OkapiPermissionSet entity, Map<String, String> okapiHeaders,
-                                    Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+      Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
     try {
       String tenantId = TenantTool.tenantId(okapiHeaders);
-      savePermList(entity.getPerms(), vertxContext, tenantId).onComplete(savePermsRes -> {
-        if (savePermsRes.failed()) {
-          String err = savePermsRes.cause().getMessage();
-          logger.error(err, savePermsRes.cause());
-          asyncResultHandler.handle(Future.succeededFuture(
-              PostTenantpermissionsResponse.respond400WithTextPlain(err)));
-          return;
-        }
-        asyncResultHandler.handle(Future.succeededFuture(
-            PostTenantpermissionsResponse.respond201WithApplicationJson(entity)));
-      });
+      handlePermLists(entity, vertxContext, tenantId)
+          .compose(v -> addMissingDefinedContext(vertxContext, tenantId))
```

Comment:
How does this work? After handling perms for **a specific module**, then update empty `definedBy` field to be `User` for all other perms (other modules') for that tenant? During bootstrap, if tenant has 3 modules enabled. Okapi sends modA perms over first, would this incorrectly mark perms from modB and modB as `User`?

Reply:
<face-palm> you're right - this was pretty short-sighted on my part.  I have to work thought some examples.

Reply:
Removed this.  See related comment threads for additional details.